pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,125 @@
+
+    **Examples**
+
+    An example of obtaining the frequencies from a set of unique eigenvals and obtaining the
+    parameter shift rule:
+
+    >>> unique_eigenvals = [1, -1, 0]
+    >>> unique_eigenvals = sorted(unique_eigenvals)
+    >>> frequencies = set()
+    >>> for i in range(len(unique_eigenvals)):
+    >>>     for j in range(i+1, len(unique_eigenvals)):
+    >>>         frequencies.add(unique_eigenvals[j] - unique_eigenvals[i])
+    >>> get_shift_rule(tuple(frequencies))
+    ([[0.8535533905932737, 1, 0.7853981633974483], [-0.14644660940672624, 1, 2.356194490192345],
+    [-0.8535533905932737, 1, -0.7853981633974483], [0.14644660940672624, 1, -2.356194490192345]],)
+
+    An example with user specified shift values:
+
+    >>> frequencies = (1, 2, 4)
+    >>> shifts = (np.pi / 3, 2 * np.pi / 3, np.pi / 4)
+    >>> get_shift_rule(frequencies, shifts)
+    ([[-2.0907702751760278, 1, 1.0471975511965976], [0.2186308015824754, 1, 2.0943951023931953],
+    [3.0000000000000004, 1, 0.7853981633974483], [2.0907702751760278, 1, -1.0471975511965976],
+    [-0.2186308015824754, 1, -2.0943951023931953], [-3.0000000000000004, 1, -0.7853981633974483]],)
+    """
+
+    n_freqs = len(frequencies)
+
+    if len(set(frequencies)) != n_freqs or any(freq <= 0 for freq in frequencies):
+        raise ValueError(
+            "Expected frequencies to be a list of unique positive values, instead got {}.".format(
+                frequencies
+            )
+        )
+
+    if shifts is None:  # assume equidistant shifts
+        mu = np.arange(1, n_freqs + 1)
+        shifts = (2 * mu - 1) * np.pi / (2 * n_freqs)
+    else:
+        shifts = list(shifts)
+        if len(shifts) != n_freqs:
+            raise ValueError(
+                "Expected number of shifts to equal the number of frequencies ({}), instead got {}.".format(
+                    n_freqs, shifts
+                )
+            )
+        if len(set(shifts)) != n_freqs:
+            raise ValueError("Shift values must be unique, instead got {}".format(shifts))
+
+    frequencies = qml.math.sort(qml.math.stack(frequencies))
+    freq_min = frequencies[0]
+
+    if np.allclose(
+        np.array(frequencies) / freq_min, range(1, len(frequencies) + 1)
+    ):  # equidistant case
+
+        mu = np.arange(1, n_freqs + 1)
+        shifts = (2 * mu - 1) * np.pi / (2 * n_freqs * freq_min)
+        coeffs = (
+            freq_min
+            * (-1) ** (mu - 1)
+            / (4 * n_freqs * np.sin(np.pi * (2 * mu - 1) / (4 * n_freqs)) ** 2)
+        )
+
+    else:  # non-equidistant case
+        sin_matr = -4 * np.sin(np.outer(shifts, frequencies))
+        sin_matr_inv = np.linalg.inv(sin_matr)
+        coeffs = -2 * np.tensordot(frequencies, sin_matr_inv, axes=1)
+
+    coeffs = np.concatenate((coeffs, [-coeff for coeff in coeffs]))
+    shifts = np.concatenate((shifts, [-shift for shift in shifts]))

```

Comment:
I think this is the last place I see for vectorization 🙂 
```suggestion
    coeffs = np.concatenate((coeffs, -coeffs))
    shifts = np.concatenate((shifts, -shifts))
```